### DISCUSSION TITLE
How to get the variable value in the former and next nodes?

### DISCUSSION THREAD

**Discussion by YuhangNIU** [idaholab/moose#18274] [Q&A General] [answered]
Hi all,

As we all know, the current variable value can be described as u(i) = _u[_qp]. Could anyone kindly tell me how to get the variable value in the former and next nodes? I want to calculate the formula, just like:
A*u(i-1) + B*u(i+1)

Thank you!
Yuhang

**Comment by WilkAndy** [comment, marked as answer]
You'll have to be more precise about what you mean by "former and next nodes".   Do you mean the next node in the list of nodes of the particular element.  Or nodes in a neighbouring element.  Or something else?

**Reply by YuhangNIU** [reply]
If we use the 1D GeneratedMesh setting nx=100, there are 100 nodes in this mesh. 
Supposing the current variable value in the fifth node is described as u(i) = _u[_qp], "the former" represents variable value in the fourth node, and "the next" represents variable value in the sixth node.

**Reply by WilkAndy** [reply]
What are you actually trying to do?  Form a spatial derivative?  MOOSE has that sort of functionality inbuilt.  Or are you doing some numerical stabilization involving values at lots of nodes?  That's more complicated, but has been done in MOOSE too, so we can help you.

The chances are that MOOSE has an elegant way of doing what you want.  BUT, if you just want all raw nodal values, you can use a NodalValueSampler VectorPostprocessor to record all the nodal values.

**Reply by YuhangNIU** [reply]
Thanks for the help. I'll try to use the NodalValueSampler! Much appreciated.

**Reply by ngrilli** [reply]
Dear @YuhangNIU 

Inside a kernel, you can access the value of the variable at the nodes using:
_u_nodal(_var.dofValues()),
this is a vector that contains the nodal values. You can see an example code here:
https://github.com/idaholab/moose/blob/next/framework/src/kernels/ConservativeAdvection.C

Best Regards,
Nicolò
